### PR TITLE
Fix #2176: InChI functions should return NULL on un-InChI-able input molecules.

### LINKED ISSUE #2176: PgSQL: mol_inchi() and mol_inchikey() should return NULL on failure.
**Description:**

Latest RDKit source as of 2018-11-24.

The cartridge mol_inchi() and mol_inchikey() functions return a blank string when the input molecule cannot be converted to InChI. I would have expected a NULL value to be returned in those cases.

```
regression=# select coalesce(mol_inchi('CC*'), '<NULL>');
 coalesce 
----------
 
(1 row)

regression=# select coalesce(mol_inchikey('CC*'), '<NULL>');
 coalesce 
----------
 
(1 row)
```
Above, I would expect '&lt;NULL&gt;' to be returned.

### PR DESCRIPTION
**Pull request by janholstjensen**
#### Reference Issue
Fixes #2176

#### What does this implement/fix? Explain your changes.
Adds check of returned string from MolInchi() and MolInchiKey() and converts blank results into NULL. Valid InChI strings or InChI keys will never be blank strings.

### CODE REVIEW

File: Code/PgSQL/rdkit/mol_op.c [hunk 1]
```diff
@@ -222,9 +222,15 @@ Datum mol_inchi(PG_FUNCTION_ARGS) {
       searchMolCache(fcinfo->flinfo->fn_extra, fcinfo->flinfo->fn_mcxt,
                      PG_GETARG_DATUM(0), NULL, &mol, NULL);
   str = MolInchi(mol, opts);
-  res = pnstrdup(str, strlen(str));
-  free((void *)str);
-  PG_RETURN_CSTRING(res);
+  if (!str || *str == 0) {
+    free((void *)str);
```
Comment (greglandrum) [inline Code/PgSQL/rdkit/mol_op.c]: If the string is null we shouldn't be freeing it.
It's better to remove this line

File: Code/PgSQL/rdkit/mol_op.c [hunk 2]
```diff
@@ -238,9 +244,15 @@ Datum mol_inchikey(PG_FUNCTION_ARGS) {
       searchMolCache(fcinfo->flinfo->fn_extra, fcinfo->flinfo->fn_mcxt,
                      PG_GETARG_DATUM(0), NULL, &mol, NULL);
   str = MolInchiKey(mol, opts);
-  res = pnstrdup(str, strlen(str));
-  free((void *)str);
-  PG_RETURN_CSTRING(res);
+  if (!str || *str == 0) {
+    free((void *)str);
```
Comment (greglandrum) [inline Code/PgSQL/rdkit/mol_op.c]: same here: don't free null string

**Comment by janholstjensen** [comment]
Fair enough. Some compilers will have undefined behavior when freeing a NULL pointer. Will be back with an updated pull request soon.

**Comment by bp-kelley** [comment]
Freeing a null pointer is guaranteed to be safe.  One shouldn’t bother checking for null, just call free regardless.

http://pubs.opengroup.org/onlinepubs/009695399/functions/free.html

Dereferencing a null pointer though....

**Comment by janholstjensen** [comment]
Branch has been updated - removed the check for NULL since the functions called will always return a non-NULL result. The only thing I need to check for is whether a blank string is returned (*str == 0).

**Comment by greglandrum** [comment]
Thanks!

**Pull request merged**